Src/ILGPU.Algorithms/ConcurrentStreamProcessor.cs

```diff
@@ -0,0 +1,161 @@
+
+        /// <summary>
+        /// Processes the given action concurrently by using the underlying accelerator
+        /// streams concurrently to submit different jobs in parallel.
+        /// </summary>
+        /// <remarks>
+        /// Note that this method assumes that all previous jobs have been synchronized
+        /// with the current processing thread.
+        /// </remarks>
+        /// <param name="numActions">The number of actions to submit.</param>
+        /// <param name="action">
+        /// The action to invoke on each stream to submit work.
+        /// </param>
+        public void ProcessConcurrently(
+            int numActions,
+            Action<AcceleratorStream, int> action) =>
+            ProcessConcurrently(null, numActions, action);
+
+        /// <summary>
+        /// Processes the given action concurrently by using the underlying accelerator
+        /// streams concurrently to submit different jobs in parallel.
+        /// </summary>
+        /// <param name="stream">The current accelerator stream.</param>
+        /// <param name="numActions">The number of actions to submit.</param>
+        /// <param name="action">
+        /// The action to invoke on each stream to submit work.
+        /// </param>
+        public void ProcessConcurrently(
+            AcceleratorStream stream,
+            int numActions,
+            Action<AcceleratorStream, int> action)
+        {
+            if (numActions < 0)
+                throw new ArgumentOutOfRangeException(nameof(numActions));
+            if (numActions == 0)
+                return;
+            if (action == null)
+                throw new ArgumentOutOfRangeException(nameof(action));
+
+            if (stream != null && numActions == 1 && MaxNumConcurrentStreams == 1)
+            {
+                // Use the given stream to process the action request
+                action(stream, 0);
+            }
+            else
+            {
+                // Wait for the current stream to finish processing
+                stream?.Synchronize();
+
+                // Perform work on all streams
+                int numActionsPerStream = XMath.DivRoundUp(
+                    numActions,
+                    MaxNumConcurrentStreams);
+                var actionStride = new Stride2D.DenseY(numActionsPerStream);
+                Parallel.For(0, MaxNumConcurrentStreams, i =>
+                {
+                    var currentStream = streams[i];
+                    using var binding = currentStream.BindScoped();
+
+                    for (int j = 0; j < numActionsPerStream; ++j)
+                    {
+                        int actionIndex = actionStride.ComputeElementIndex((i, j));
+                        if (actionIndex >= numActions)
+                            break;
+                        action(currentStream, actionIndex);
+                    }
+
+                    // Wait for the result
+                    currentStream.Synchronize();
```

Comment:
Again, is a synchronization necessary?  Can you just queue jobs and only synchronize later if you need to?

Reply:
Unfortunately, not at the moment as ILGPU does not feature a way to attach streams to each other.